returnn/torch/engine.py

```diff
@@ -259,6 +259,50 @@ def _load_model(self, epoch):
             print("Load model %s" % (filename,), file=log.v4)
             model_state = torch.load(filename)
             self._model.load_state_dict(model_state)
+        preload_from_files = self.config.typed_value("preload_from_files", {})
+        if preload_from_files:
+            # see `preload_from_files` in tf engine and `returnn.tf.network.CustomCheckpointLoader`
+            is_training = self.config.value("task", "train") == "train"
+            is_first_train_epoch = epoch == 1 and (
+                is_training or self.config.value("task", "train") == "initialize_model"
+            )
+            # We use the reversed sorted order here to achieve consistent behavior with the TF engine.
+            # There, the keys are used in sorted order but if a variable is loaded,
+            # it will not be considered anymore afterwards.
+            # So the first occurrence is used.
+            # Here, we overwrite variables even if they have been loaded before.
+            # In order to get consistent behavior, we use the reversed order.
+            for preload_key, opts in reversed(sorted(preload_from_files.items())):
+                assert isinstance(opts, dict) and "filename" in opts
+                if opts.get("init_for_train", False):
+                    if not is_first_train_epoch:
+                        continue
+                else:  # default: init for recog
+                    if is_training:
+                        continue
+                print(f"Pre-load weights for key '{preload_key}' from {opts['filename']}", file=log.v3)
+                preload_model_state = torch.load(opts["filename"])
+                if opts.get("checkpoint_key", None) is not None:
```

Comment:
Why this new option? Do some external models come with checkpoints that do not only store the state dict but also other things? I'm not sure we have to try to handle that. Because it could be in a tuple, or in a nested dict. I would rather require the user to convert the checkpoint if necessary such that it looks like it has been saved via `torch.save(model.state_dict())`. Or is there some other convention I'm missing? This is not for `torch.save(model)`, right?

Reply:
Good point. I tried importing a wav2vec 2.0 checkpoint from [here](https://github.com/facebookresearch/fairseq/blob/main/examples/wav2vec/README.md). When I `torch.load()` it, it is a dict with `dict_keys(['args', 'model', 'optimizer_history', 'extra_state', 'last_optimizer_state'])`. Since my goal was to load this model and the option requires minimal implementation, I added it. I'm not sure if there are conventions, though.

Reply:
I think also here it would help maybe to add this example with the w2v checkpoint as a comment.

Reply:
> Shouldn't we maybe also follow more this standard convention?

It's farfetched to call anything a convention here. But true, many frameworks seem to add additional information into the checkpoint files and it's also in the docs https://pytorch.org/tutorials/beginner/saving_loading_models.html#save

```
torch.save({
            'epoch': epoch,
            'model_state_dict': model.state_dict(),
            'optimizer_state_dict': optimizer.state_dict(),
            'loss': loss,
            ...
            }, PATH)
```
(Note the dots...)

@Icemole implemented saving the optimizer state to separate files. I think it can be convenient this way because you directly get the checkpoint you need for inference. But of course, if we needed to save more things then there is a limit to this approach.

> We can also extend this option, to accept a list of keys, so it can go deeply through dict keys, list items, etc.

Yeah, and allow it to be a callable so really everything is possible... :ok_woman:

I would rather have a defined interface and just support loading what we save. And yes, maybe this `checkpoint_key` for 1-level dicts because it's common.

Reply:
I added a comment. Is this thread resolved then?